### ISSUE TITLE
Reduce texture upload overhead

### ISSUE THREAD

**Issue by Michael-Lfx** [gfx-rs/gfx#2738]
https://github.com/gfx-rs/gfx/blob/b5163cc21e8500936635a11d99912ca5b91ad0de/examples/quad/main.rs#L309

https://github.com/gfx-rs/gfx/blob/b5163cc21e8500936635a11d99912ca5b91ad0de/examples/quad/main.rs#L509

Is it possible that replacing OneShot with MultiShot command buffer for uploading data to the same texture repeatedly just like @termhn introduced triple-buffering in quad example? cc @kvark

**Comment by termhn** [comment]
The texture only gets uploaded once though I believe...

**Comment by Michael-Lfx** [comment]
> The texture only gets uploaded once though I believe...
> […](#)

Situation like `glTexImage2D` could upload multiple times whenever user calls. Is it alright to use `command::MultiShot` to solve the problem?

**Comment by msiglreith** [comment]
yes, reusing command buffers is fine (if the backend supports it so far). It helps reducing the overhead of re-encoding the commands. Src and Dst texture will be the same ofc then and can't be changed afterwards.

**Comment by Michael-Lfx** [comment]
I will change the src and dst when implement glTexImage2D on top of hal. So the answer is yes to the question? @msiglreith

**Comment by msiglreith** [comment]
Hmm, rather no, I don't think there is much room for optimizations via reusable command buffers for this use case I fear. They help to reduce CPU load for reptetitive tasks (for example identical work over multiple frames). OpenGL API doesn't provide an easy way to make use of this I fear.

**Comment by Michael-Lfx** [comment]
EDIT: I will NOT change the dst when implementing glTexImage2D on top of hal. The dst is the struct itself to hold the hal Image and ImageView. I think the src may change as each call will the last parameter of glTexImage2D to provide a new data source.

**Comment by kvark** [comment]
Is this still an issue? I'm having a hard time understanding what the problem is, tbh. Please elaborate on the use case if it's still a problem.

**Comment by Michael-Lfx** [comment]
Yes. it's our usage issues but not a hal issue. I want to use one permanent `MultiShot command` and a staging buffer for all the future data transferring if it' permitted to do so when implementing `glTexImage2D()`.

**Comment by kvark** [comment]
I don't think this is going to save you much versus just recording a `OneShot` command buffer whenever you need. It feels like a premature optimization at this point.

Overall, I'd expect a first version of GL implementation to just keep a single one-shot command buffer recording at a time, which gets submitted at flush points (i.e. `glFlush`, `glFinish` and swap buffers).

**Comment by Michael-Lfx** [comment]
> TODO: we can't keep a pointer to the user data anywhere outside of the implementation of `glTex*Image*` calls. Therefore, we shouldn't try to store it somewhere. Instead, we need to use it right away. For texture uploads, this means allocating a piece of staging buffer memory:
> - one with CPU_VISIBLE property
> - doesn't have to be a separate allocation or a new buffer
> - can just be an unused part of an existing buffer
> Then copying the data to this memory (so that the pointer isn't needed any more) and issuing a `copy_buffer_to_texture` operation on a currently recording command buffer to copy over these staging contents into the actual GPU memory of a texture layer/level.

I have succeed begin a `CommandBuffer` inside `eglSwapBuffers()` and `eglMakeCurrent()` and then execute all the rest operations on it on a user FBO to system FBO scenario. And it works.

But I get a nonsense(trash data) image showing in the screen when I try to issue `copy_buffer_to_image()` without calling `CommandBuffer::finish()` and `queue::submit_without_semaphores` on the active `CommandBuffer`.

**Comment by kvark** [comment]
Please run with Vulkan Validation to see if you are missing any barriers between transfers and using the image.
